lib/SPIRV/SPIRVWriter.cpp

```diff
@@ -2227,6 +2228,77 @@ SPIRVValue *LLVMToSPIRV::transIntrinsicInst(IntrinsicInst *II,
                                 transValue(II->getArgOperand(2), BB), BB);
     return applyRoundingModeConstraint(II->getOperand(3), BI);
   }
+  case Intrinsic::fshl: {
+    // We will implement fshl(a, b, c) in SPIR-V instructions.
+    // "Funnel shift rotate" operates on same-sized integers; its return type
+    // also matches.
+    SPIRVType *Ty = transType(II->getType());
+    // Create a separate function - otherwise we'd have to rework the CFG
+    // of the current function. Then simply translate the intrinsic call
+    // into the new function.
+    auto *FSHLFuncTy = BM->addFunctionType(Ty, {Ty, Ty, Ty});
+    auto *FSHLFunc = BM->addFunction(FSHLFuncTy);
```

Comment:
I suggest to add a name to this function and query it first to check if it exists already (might be the case if there are several calls to `fshl`). For ideas how to "mangle" function name you can take a look at https://github.com/KhronosGroup/SPIRV-LLVM-Translator/pull/614/files#diff-7990d99acb2f8325d245dfecdab758eae5aebae2b54c04fd8e906212eb80ae10R2438

Also, by having specially-defined function name for `fshl` replacement we could replace call to that function back to `fshl` intrinsic in SPIR-V Reader (this is an idea for future improvement - not asking to do this right now)

Reply:
Is it a good enough practice to query such functions by name instead of, say, hashing the functions by integer bitsize? The latter wouldn't interfere with naming the functions properly, so it's merely a question of whether names should be used for such queries.

Reply:
> Is it a good enough practice to query such functions by name instead of, say, hashing the functions by integer bitsize?

I can't say exactly whether it is the best thing from the performance prospective, but this (querying functions by name) is at least widely-used mechanism:

https://github.com/KhronosGroup/SPIRV-LLVM-Translator/blob/e1405604253a8569375c8031a73e020af2da9bd1/lib/SPIRV/OCLUtil.h#L523-L525

The snippet above is called from 5 different places in the translator. Also you can find 6 more direct uses of [`llvm::Module::getOrInsertFunction`](https://llvm.org/doxygen/classllvm_1_1Module.html#abb107e4edbf05eae92936cba6801c2d9) like:

https://github.com/KhronosGroup/SPIRV-LLVM-Translator/blob/e1405604253a8569375c8031a73e020af2da9bd1/lib/SPIRV/SPIRVReader.cpp#L1453

Personally, I like the idea of querying by names more because of its scalability: if you decide to hash functions by integer bitsize, then you need to have a separate map for each function you would like to hash and I believe that we will have more unsupported intrinsics which we will replace with corresponding implementation in the translator.

Also, having predefined names allows to do some optimizations like:

> we could replace call to that function back to fshl intrinsic in SPIR-V Reader 

An example of such would be:

https://github.com/KhronosGroup/SPIRV-LLVM-Translator/blob/e1405604253a8569375c8031a73e020af2da9bd1/lib/SPIRV/OCLUtil.cpp#L1557-L1562

> The latter wouldn't interfere with naming the functions properly, so it's merely a question of whether names should be used for such queries.

We usually use something like this:

https://github.com/KhronosGroup/SPIRV-LLVM-Translator/blob/e1405604253a8569375c8031a73e020af2da9bd1/lib/SPIRV/SPIRVInternal.h#L356-L360

However, querying whether function exists in `SPIRVModule` will require updating existing infrastructure: right now functions are just stored in a vector. Such query is much easier to do at LLVM IR level and as I can see, all functions are stored within LLVM `Module` using `StringMap` (`StringRef` -> `Value`) - you might want to consider moving this piece of code into `SPIRVRegularize` pass

Reply:
As a matter of fact, couldn't I just do `FSHLIntrinsicPtr->getDeclaration()->replaceAllUsesWith(ImplementedFSHLPtr)`? This definitely turns the tables in `SPIRVRegularizeLLVM`'s favor.

Reply:
Done, thanks for the detailed suggestion!
Should I also add a test for function re-usage scenario/another integer type?

Reply:
> As a matter of fact, couldn't I just do FSHLIntrinsicPtr->getDeclaration()->replaceAllUsesWith(ImplementedFSHLPtr)?
> UPD: This exact trick wouldn't work for call instructions, but I'm still going to try out your suggestions.

I'm not sure that this is absolutely correct to replace all uses of declaration with another declaration. However, replacing all uses of particular `CallInst` with another `Value` (like another `CallInst)` should work fine

Reply:
> I'm not sure that this is absolutely correct to replace all uses of declaration with another declaration. However, replacing all uses of particular CallInst with another Value (like another CallInst) should work fine

I've ended up mimicking `lowerMemset()`'s behavior, i.e. using `replaceCalledFunction()`

Reply:
> Should I also add a test for function re-usage scenario/another integer type?

Yes, please. I think both tests (re-usage of existing type and using another integer type) are very useful as we can encounter such use-cases in real apps

Reply:
Done.